src/Search/Services/SearchableService.php

```diff
@@ -0,0 +1,162 @@
+ */
+class SearchableService
+{
+    use Injectable;
+    use Extensible;
+    use Configurable;
+
+    /**
+     * Skip the canView() check at a class level to increase performance of search reindex.
+     * Be careful as this may lead to content showing in search results that should not be there such as non-public,
+     * cms-user-only content.  This may potentially happen via edge cases such as skipping checks where subclasses
+     * are involved.
+     *
+     * This has no effect on when search results as canView() must still be run there
+     *
+     * @var array namespaced classes to skip canView() check on search reindex
+     */
+    private static $classes_that_skip_indexing_canview_check = [];
+
+    /**
+     * Non-persistant memory cache that only lasts the lifetime of the request
+     *
+     * @var array
+     */
+    private $cache = [];
+
+    /**
+     * Clears the internal cache
+     */
+    public function clearCache(): void
+    {
+        $this->cache = [];
+    }
+
+    /**
+     * Used during search reindex
+     *
+     * This is considered the primary layer of protection
+     *
+     * @param DataObject $obj
+     * @return bool
+     */
+    public function isIndexable(DataObject $obj): bool
+    {
+        return $this->isSearchable($obj, true);
+    }
+
+    /**
+     * Used when retrieving search results
+     *
+     * This is considered the secondary layer of protection
+     *
+     * It's important to still have this layer in conjuction with the index layer as non-searchable results may be
+     * in the search index because:
+     * a) they were added to the index pre-fulltextsearch 3.7 and a reindex to purge old records was never run, OR
+     * b) the DataObject has a non-deterministic canView() check such as `return $date <= $dateOfIndex;`
+     *
+     * @param DataObject $obj
+     * @return bool
+     */
+    public function isViewable(DataObject $obj): bool
+    {
+        return $this->isSearchable($obj, false);
+    }
+
+    /**
+     * Checks and caches whether the given DataObject can be indexed. This is determined by two factors:
+     * - Whether the ShowInSearch property / getShowInSearch() method evaluates to true
+     * - Whether the canView method evaluates to true against an anonymous user (optional, can be disabled)
+     *
+     * @param DataObject $obj
+     * @return bool
+     */
+    private function isSearchable(DataObject $obj, bool $indexing): bool
+    {
+        // check if is a valid DataObject that has been persisted to the database
+        if (is_null($obj) || !$obj->ID) {
+            return false;
+        }
+
+        $key = $this->getCacheKey($obj);
+        if (isset($this->cache[$key])) {
+            return $this->cache[$key];
+        }
+
+        $value = true;
+
+        // ShowInSearch check
+        // This will also call $obj->getShowInSearch() if it exists
+        if (isset($obj->ShowInSearch) && !$obj->ShowInSearch) {
+            $value = false;
```

Comment:
> ShowInSearch=false, since they also need to check canView()

That's incorrect, the canView() block below has a `if ($value)` check before doing the canView() check

ShowInSearch is inexpensive, and canView() could be potentially expensive

In terms of extensible performance, you could disable the canView check on a class via configuration and then do the canView check in your extension classes.  This will prevent an uncached duplicate canView() check so performance shouldn't be impacted.

Reply:
Right of course, I missed the `if ($value)`